File: nlox/Lox/Stmt.cs
abstract record Stmt();

record ExpressionStmt(Expr Expression) : Stmt;
record Print(Expr Expression) : Stmt;
record Var(Token Name, Expr? Initializer) : Stmt;
record Block(List<Stmt> Statements) : Stmt;
record If(Expr Condition, Stmt ThenBranch, Stmt? ElseBranch) : Stmt;
record While(Expr Condition, Stmt Body) : Stmt;
record Function(Token Name, List<Token> Params, List<Stmt> Body) : Stmt;
record Return(Token Keyword, Expr? Value) : Stmt;
record ClassDecl(Token Name, Variable? Superclass, List<Function> Methods): Stmt;

File: tslox/src/stmt.ts
import { Expr, Variable } from "./expr"
import { Token } from "./token"

export type If = {
  kind: "If",
  condition: Expr,
  thenBranch: Stmt,
  elseBranch: Stmt | null
}

export type Print = {
  kind: "Print",
  expression: Expr
}

export type Return = {
  kind: "Return",
  keyword: Token,
  value: Expr | null
}

export type While = {
  kind: "While",
  condition: Expr,
  body: Stmt
}

export type Block = {
  kind: "Block",
  statements: Stmt[]
}

export type ClassDecl = {
  kind: "ClassDecl",
  name: Token,
  superclass: Variable | null,
  methods: Function[]
}

export type Function = {
  kind: "Function",
  name: Token,
  params: Token[],
  body: Stmt[]
}

export type Var = {
  kind: "Var",
  name: Token,
  initializer: Expr | null
}

export type ExpressionStmt = {
  kind: "ExpressionStmt",
  expression: Expr
}

export type Stmt =
    If
  | Print
  | Return
  | While
  | Block
  | ClassDecl
  | Function
  | Var
  | ExpressionStmt;
